### PR TITLE
fix(pacquet): base repeat-install freshness on filesystem mtime at nanosecond precision

### PR DESCRIPTION
**Pull request by zkochan**
## Summary

The optimistic repeat-install fast path (and `verify-deps-before-run`) records `lastValidatedTimestamp` and later compares it against manifest / `.pnpmfile.cjs` / patch mtimes to decide whether anything changed. Pacquet compared a **wall-clock** timestamp against **filesystem** mtimes, which fails in two independent ways — both surfaced as flaky pacquet-CLI test failures once the Rust test job moved onto Blacksmith's Linux runners:

1. **Wall-vs-filesystem clock skew.** `lastValidatedTimestamp` came from `now_millis()` (wall clock). On Blacksmith's Linux microVM the wall clock runs **~2–3 ms ahead** of the filesystem's mtime clock (measured), so a file edited a couple ms after an install gets an mtime *below* the baseline and the strict `mtime > baseline` comparison misses the edit.
2. **Same-millisecond collision.** Pacquet installs are fast enough that the lockfile one install writes and a manifest edited immediately after can land in the **same millisecond**; the comparison truncated both to whole ms, so `manifest.ms == lockfile.ms` read as "unchanged".

### Fix (two commits, pacquet-only)

- **Base the freshness timestamp on filesystem mtimes** — the max of the lockfile this install wrote and the manifests it validated — instead of the wall clock, so the baseline and the mtimes it's compared against share one clock (no skew). The max preserves the content-check-pass optimization (an already-edited but validated manifest is still blessed). This mirrors pnpm's `checkDepsStatus`, whose single-project path already keys off the lockfile mtime; pacquet had diverged.
- **Compare at nanosecond precision.** The recorded timestamp stays millisecond-precise (a rounded-down mtime, kept ms for cross-stack parity), but the subject is compared in nanoseconds, so a file whose whole-ms mtime ties the baseline but was written later in that ms is still seen as modified.

### Verification

- **On the actual skewed Blacksmith Linux runner** (probe confirmed 2–3 ms skew still present): the previously-flaky tests ran **40/40 green**, and the full pacquet suite (5380 tests) passed. The skew fix alone left a 2/40 residual — the same-ms flake — which the nanosecond fix closes.
- **Deterministic local proof:** a forced same-millisecond / higher-nanosecond edit passes only with the nanosecond comparison and fails with the millisecond one.
- 709 unit tests + the CLI integration suites pass locally; fmt / clippy / doc / dylint clean.

Note: this is the actual root cause of the Blacksmith pacquet flakiness. An earlier fix (pnpm/pnpm#12975) addressed whole-second mtime *granularity*, which Blacksmith doesn't have — it's harmless (and still helps genuinely coarse filesystems) but was a misdiagnosis; happy to revert it separately if you'd prefer.

## Squash Commit Body

```text
Pacquet's repeat-install fast path recorded lastValidatedTimestamp from
the wall clock but compared it against filesystem mtimes. On a runner
whose wall clock runs ahead of the filesystem's mtime clock (~2-3 ms on
Blacksmith Linux), a file edited just after an install got an mtime below
the baseline and the edit was missed; and because installs are fast, the
lockfile and a same-instant edit could share a millisecond, which the
ms-truncated comparison also missed.

Record the baseline from filesystem mtimes (max of the lockfile and the
validated manifests) so it shares a clock with the files it is compared
against, and compare the subject at nanosecond precision against the
rounded-down millisecond reference. Mirrors pnpm's checkDepsStatus, which
already keys single-project freshness off the lockfile mtime.
```

## Checklist

- [x] The change is implemented in both the TypeScript CLI and the Rust `pnpm/` port, or the description notes what still needs porting. _(Pacquet-only: it aligns pacquet with pnpm's existing single-project behavior; no TS change needed.)_
- [x] Added a changeset. _(`pacquet`, patch.)_
- [x] Added or updated tests. _(Nanosecond-precision unit test; the CLI integration tests are the regression coverage — verified 40/40 on the skewed runner.)_
- [x] Updated the documentation if needed. _(N/A.)_

---
Written by an agent (Claude Code, claude-opus-4-8).

---
<a href="https://app.blacksmith.sh/pnpm/codesmith/pnpm/pr/12981"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with Codesmith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1786555898&installation_id=145934176&pr_number=12981&repository=pnpm%2Fpnpm&return_to=https%3A%2F%2Fgithub.com%2Fpnpm%2Fpnpm%2Fpull%2F12981&signature=5ed8e3ce21c913053907e72c220108c50cf318f8d0a0b894d4d2e6b235bf78cb"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with Codesmith</picture></a>
<sup>Need help on this PR? Tag <code>/codesmith</code> with what you need. Autofix is disabled.</sup>

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed optimistic repeat-install “already up to date” detection after quick edits to project manifests, patch files, or lockfile-related inputs.
  * Improved freshness checks by using higher-precision file modification timestamps, ensuring changes within the same millisecond are detected.
  * Updated lockfile freshness evaluation to avoid incorrect re-resolution skips and reduce timestamp-resolution mismatches.
* **Tests**
  * Added and updated regression tests for nanosecond-precision behavior and lockfile baseline edge cases.

---

### Update (review round)

CodeRabbit and Qodo both flagged a real regression in the first cut: the nanosecond comparison made the **unchanged lockfile self-flag** against its own millisecond-truncated baseline, forcing a content check on every repeat install. Fixed in `82b4c352dd` — the lockfile freshness check now compares at whole-millisecond precision (the unchanged lockfile ties its baseline; an external edit lands in a later ms and is still caught), while manifests/patches/pnpmfiles keep nanosecond precision (needed for same-ms edits, and can't self-flag because they compare against a *different* file). Added `lockfile_check_does_not_self_flag_its_own_baseline`. Re-verified **40/40 green** on the skewed Blacksmith Linux runner.

Known residual (documented for reviewers): after a **workspace** content-check pass on a manifest edited with unchanged deps, that manifest becomes the stored baseline and self-flags on later runs — a content check repeats until a real re-resolve. This matches pnpm's *single-project* behavior; pnpm's *workspace* path avoids it via a wall-clock refresh. Fully closing it needs a larger workspace-state change (it collides with the workspace unit tests and pnpm's ms state-format), so it's left as follow-up. The common fully-unchanged repeat install is fast again.

### CODE REVIEW

File: pnpm/crates/package-manager/src/optimistic_repeat_install.rs [hunk 1]
```diff
@@ -874,18 +888,51 @@ fn mtime_ms(path: &Path) -> Option<i64> {
...
-/// common case is unaffected.
+/// a needed install.
 fn modified_at_or_after(subject: FileMtime, reference_ms: i64) -> bool {
-    if subject.whole_second { subject.ms + 1_000 > reference_ms } else { subject.ms > reference_ms }
+    let reference_ns = reference_ms.saturating_mul(NANOS_PER_MILLI);
+    if subject.whole_second {
+        subject.ns.saturating_add(NANOS_PER_SEC) > reference_ns
+    } else {
+        subject.ns > reference_ns
+    }
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm/crates/package-manager/src/optimistic_repeat_install.rs:906]: Good catch — fixed in 82b4c352dd. The nanosecond comparison is now used only for manifests/patches/pnpmfiles (which need it for same-millisecond edits and are compared against a *different* file — the lockfile — so they can't self-flag). The lockfile's own freshness check (`wanted_lockfile_modified`) now compares at whole-millisecond precision, so the unchanged lockfile ties its own millisecond-truncated baseline instead of reporting "modified" every repeat install; an external lockfile edit lands in a later millisecond and is still caught. Added `lockfile_check_does_not_self_flag_its_own_baseline` covering the unchanged-baseline-source case.

---
Written by an agent (Claude Code, claude-opus-4-8).

File: pnpm/crates/package-manager/src/optimistic_repeat_install/tests.rs [hunk 2]
```diff
@@ -2225,24 +2225,30 @@ fn does_not_regenerate_wanted_lockfile_when_lockfile_writing_disabled() {
...
+    let on_boundary = FileMtime { ms, ns, whole_second: false };
+    assert!(!modified_at_or_after(on_boundary, ms));
+    assert!(!modified_at_or_after(on_boundary, ms + 1));
+    assert!(modified_at_or_after(on_boundary, ms - 1));
+
+    // Same millisecond as the reference, half a millisecond later: the
+    // millisecond values tie, but the nanosecond mtime does not, so the
+    // edit is still seen (the same-millisecond flake this guards against).
+    let later_in_same_ms = FileMtime { ms, ns: ns + 500_000, whole_second: false };
+    assert!(modified_at_or_after(later_in_same_ms, ms));
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm/crates/package-manager/src/optimistic_repeat_install/tests.rs:2254]: Good catch — fixed in 82b4c352dd. The nanosecond comparison is now used only for manifests/patches/pnpmfiles (which need it for same-millisecond edits and are compared against a *different* file — the lockfile — so they can't self-flag). The lockfile's own freshness check (`wanted_lockfile_modified`) now compares at whole-millisecond precision, so the unchanged lockfile ties its own millisecond-truncated baseline instead of reporting "modified" every repeat install; an external lockfile edit lands in a later millisecond and is still caught. Added `lockfile_check_does_not_self_flag_its_own_baseline` covering the unchanged-baseline-source case.

---
Written by an agent (Claude Code, claude-opus-4-8).

File: pnpm/crates/package-manager/src/optimistic_repeat_install.rs [hunk 3]
```diff
@@ -874,18 +888,51 @@ fn mtime_ms(path: &Path) -> Option<i64> {
...
+    config: &Config,
+    project_manifests: &[(PathBuf, &PackageManifest)],
+) -> Option<i64> {
+    let lockfile = mtime_ms(&workspace_root.join(Lockfile::FILE_NAME))
+        .or_else(|| mtime_ms(&config.virtual_store_dir.join(Lockfile::CURRENT_FILE_NAME)));
+    project_manifests
+        .iter()
+        .filter_map(|(_, manifest)| mtime_ms(manifest.path()))
+        .chain(lockfile)
+        .max()
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm/crates/package-manager/src/optimistic_repeat_install.rs:935]: Thanks. On the two points:

- **Wall-clock fallback:** `build_workspace_state` only falls back to `now_millis()` when neither the wanted nor the current lockfile can be stat'd — i.e. a `lockfile: false` / dependency-less install where there is no lockfile to key off. In that state there is nothing to compare against anyway, and the value is only ever a lower bound for the *next* run's freshness check, so it can't cause a needed install to be skipped.
- **Future-dated mtimes:** a file whose mtime is ahead of the baseline reads as *modified* (`mtime > baseline`), so it triggers the authoritative content check rather than skipping it — the fast path errs toward doing more work, not less. So an implausibly-future mtime is already fail-safe here.

---
Written by an agent (Claude Code, claude-opus-4-8).

File: pnpm/crates/package-manager/src/optimistic_repeat_install.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm/crates/package-manager/src/optimistic_repeat_install.rs:906]: Good catch — fixed in 82b4c352dd. The nanosecond comparison is now used only for manifests/patches/pnpmfiles (which need it for same-millisecond edits and are compared against a *different* file — the lockfile — so they can't self-flag). The lockfile's own freshness check (`wanted_lockfile_modified`) now compares at whole-millisecond precision, so the unchanged lockfile ties its own millisecond-truncated baseline instead of reporting "modified" every repeat install; an external lockfile edit lands in a later millisecond and is still caught. Added `lockfile_check_does_not_self_flag_its_own_baseline` covering the unchanged-baseline-source case.

---
Written by an agent (Claude Code, claude-opus-4-8).

File: .changeset/repeat-install-clock-skew.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+---
+"pacquet": patch
+---
```
[the remark this answers is not in the document]
Comment (zkochan) [reply .changeset/repeat-install-clock-skew.md:3]: This is a false positive: the change is Rust-only (pacquet), and in this repo's changesets `pnpm` refers specifically to the TypeScript CLI package, which this PR does not touch. Per the repo's Rust-changeset policy, a pacquet-only change targets `pacquet` and must *not* add a `pnpm` entry. Leaving it as-is.

---
Written by an agent (Claude Code, claude-opus-4-8).

File: pnpm/crates/package-manager/src/optimistic_repeat_install.rs [hunk 5]
```diff
@@ -874,28 +888,80 @@ fn mtime_ms(path: &Path) -> Option<i64> {
...
+}
+
+/// Whether a lockfile with mtime `subject_ms` post-dates `reference_ms`,
+/// at whole-millisecond precision. Unlike the nanosecond
+/// [`modified_at_or_after`] used for manifests/patches/pnpmfiles: the
+/// baseline is the lockfile's own mtime truncated to milliseconds, so a
+/// nanosecond comparison would flag the unchanged lockfile against its own
+/// truncated value every repeat install. See [`wanted_lockfile_modified`].
+fn lockfile_mtime_after(subject_ms: i64, reference_ms: i64) -> bool {
+    subject_ms > reference_ms
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm/crates/package-manager/src/optimistic_repeat_install.rs]: Good catch — fixed in the latest commit. `lockfile_modified_since` now restores the whole-second handling for the coarse-filesystem case: on a whole-second-mtime filesystem a lockfile externally edited in the same second as the install keeps an identical recorded mtime, so it's indistinguishable from the install's own write and now falls through to the content check (`ms + 1000 > reference` when `whole_second`). On a sub-second filesystem the whole-second branch never fires, so the strict millisecond comparison — and its no-self-flag property, which is the whole point on the Blacksmith Linux runner — is unchanged. Extended the unit test with the whole-second case.

The remaining sub-millisecond gap (a same-*millisecond*, sub-nanosecond-later external lockfile edit on a sub-second filesystem) is fundamentally undetectable without storing the lockfile's full-nanosecond mtime in the workspace state, and would require a same-millisecond external rewrite — not a realistic scenario for a human git checkout / manual edit, which land whole milliseconds later.

---
Written by an agent (Claude Code, claude-opus-4-8).
